src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/DefaultRazorSemanticTokenInfoService.cs

```diff
@@ -0,0 +1,213 @@
+                    {
+                        result.Add(tagHelperAttributeTag.Name);
+                    }
+                    break;
+                default:
+                    break;
+            }
+
+            var children = syntaxNode.ChildNodes();
+            foreach (var child in children)
+            {
+                result.AddRange(VisitNode(child));
+            }
+
+            return result;
+        }
+
+        private static SemanticTokens ConvertSyntaxTokensToSemanticTokens(
+            IEnumerable<SyntaxNode> syntaxTokens,
+            RazorCodeDocument razorCodeDocument,
+            SemanticTokenLegend semanticTokensLegend)
+        {
+            SyntaxNode previousToken = null;
+
+            var data = new List<uint>();
+            foreach (var token in syntaxTokens)
+            {
+                var newData = GetData(token, previousToken, razorCodeDocument, semanticTokensLegend);
+                data.AddRange(newData);
+
+                previousToken = token;
+            }
+
+            return new SemanticTokens
+            {
+                Data = data
+            };
+        }
+
+        /**
+         * In short, each token takes 5 integers to represent, so a specific token `i` in the file consists of the following array indices:
+         *  - at index `5*i`   - `deltaLine`: token line number, relative to the previous token
+         *  - at index `5*i+1` - `deltaStart`: token start character, relative to the previous token (relative to 0 or the previous token's start if they are on the same line)
+         *  - at index `5*i+2` - `length`: the length of the token. A token cannot be multiline.
+         *  - at index `5*i+3` - `tokenType`: will be looked up in `SemanticTokensLegend.tokenTypes`
+         *  - at index `5*i+4` - `tokenModifiers`: each set bit will be looked up in `SemanticTokensLegend.tokenModifiers`
+        **/
+        private static IEnumerable<uint> GetData(
+            SyntaxNode currentNode,
+            SyntaxNode previousNode,
+            RazorCodeDocument razorCodeDocument,
+            SemanticTokenLegend legend)
+        {
+            var previousRange = previousNode?.GetRange(razorCodeDocument);
+            var currentRange = currentNode.GetRange(razorCodeDocument);
+
+            // deltaLine
+            var previousLineIndex = previousNode == null ? 0 : previousRange.Start.Line;
+            yield return (uint)(currentRange.Start.Line - previousLineIndex);
+
+            // deltaStart
+            if (previousRange != null && previousRange?.Start.Line == currentRange.Start.Line)
+            {
+                yield return (uint)(currentRange.Start.Character - previousRange.Start.Character);
+            }
+            else
+            {
+                yield return (uint)(currentRange.Start.Character);
+            }
+
+            // length
+            if(currentNode.Span.Length <= 0)
+            {
+                throw new NotImplementedException();
```

Comment:
Do we need this check at all?

Reply:
That was a guard I added because I can't do math and made a silly mistake.

Reply:
Debug.Assert

Reply:
> Debug.Assert

`Debug.Fail`